x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/SeparatedValuesLogFileStructureFinder.java

```diff
@@ -0,0 +1,458 @@
+                .setTimestampFormats(timeField.v2().dateFormats)
+                .setNeedClientTimezone(timeField.v2().hasTimezoneDependentParsing())
+                .setMultilineStartPattern(timeLineRegex);
+        }
+
+        SortedMap<String, Object> mappings = guessMappings(explanation, sampleRecords);
+        mappings.put(DEFAULT_TIMESTAMP_FIELD, Collections.singletonMap(MAPPING_TYPE_SETTING, "date"));
+
+        structure = structureBuilder
+            .setMappings(mappings)
+            .setExplanation(explanation)
+            .build();
+    }
+
+    @Override
+    public List<String> getSampleMessages() {
+        return Collections.unmodifiableList(sampleMessages);
+    }
+
+    @Override
+    public LogFileStructure getStructure() {
+        return structure;
+    }
+
+    static Tuple<List<List<String>>, List<Integer>> readRows(String sample, CsvPreference csvPreference) throws IOException {
+
+        int fieldsInFirstRow = -1;
+
+        List<List<String>> rows = new ArrayList<>();
+        List<Integer> lineNumbers = new ArrayList<>();
+
+        try (CsvListReader csvReader = new CsvListReader(new StringReader(sample), csvPreference)) {
+
+            try {
+                List<String> row;
+                while ((row = csvReader.read()) != null) {
+                    if (fieldsInFirstRow < 0) {
+                        fieldsInFirstRow = row.size();
+                    } else {
+                        // Tolerate extra columns if and only if they're empty
+                        while (row.size() > fieldsInFirstRow && row.get(row.size() - 1) == null) {
+                            row.remove(row.size() - 1);
+                        }
+                    }
+                    rows.add(row);
+                    lineNumbers.add(csvReader.getLineNumber());
+                }
+            } catch (SuperCsvException e) {
+                // Tolerate an incomplete last row
+                if (notUnexpectedEndOfFile(e)) {
+                    throw e;
+                }
+            }
+        }
+
+        assert rows.isEmpty() == false;
+        assert lineNumbers.size() == rows.size();
+
+        if (rows.get(0).size() != rows.get(rows.size() - 1).size()) {
+            rows.remove(rows.size() - 1);
+            lineNumbers.remove(lineNumbers.size() - 1);
+        }
+
+        // This should have been enforced by canCreateFromSample()
+        assert rows.size() > 1;
+
+        return new Tuple<>(rows, lineNumbers);
+    }
+
+    static Tuple<Boolean, String[]> findHeaderFromSample(List<String> explanation, List<List<String>> rows) {
+
+        assert rows.isEmpty() == false;
+
+        List<String> firstRow = rows.get(0);
+        boolean isCsvHeaderInFile = true;
+
+        if (rows.size() < 3) {
+            explanation.add("Too little data to accurately assess whether header is in sample - guessing it is");
+        } else {
+            isCsvHeaderInFile = isFirstRowUnusual(explanation, rows);
```

Comment:
So, we want to automatically determine if a CSV header is given in the sample or not. Here are a couple of quick checks that we may want to try before doing expensive statistical checking and Levenshtein distances. Of course, this depends on how flexible we want to be. 

- Are all entries unique in the header row? You should not have duplicate header entries
- Do we want to support `null` headers fields? If not, it stands to reason that none should be null in the sample
- Are all entries actual `string` values? I am not sure it is valid to have a header that is just a number or a blank string.

Reply:
The duplicates test is added in https://github.com/elastic/elasticsearch/pull/32788/commits/cf2d6750c29069b358dac7e30702e99a52a73eae